services/src/users/postgres_userdb.rs

```diff
@@ -0,0 +1,175 @@
+                        "\
+                        CREATE TABLE version (\
+                            version INT\
+                        );\
+                        INSERT INTO version VALUES (1);\
+
+                        CREATE TABLE users (
+                            id UUID PRIMARY KEY,
+                            email character varying (256) UNIQUE,
+                            password_hash character varying (256),
+                            real_name character varying (256)
+                        );
+                        ",
+                    )
+                    .await
+                    .context(error::TokioPostgres)?;
+                    version += 1;
+                    // TODO log
+                    println!("Updated user database to schema version {}", version);
+                }
+                1 => {
+                    conn.batch_execute(
+                        "\
+                        ALTER TABLE users ADD COLUMN active boolean;
+
+                        UPDATE version SET version = 2;\
+                        ",
+                    )
+                    .await
+                    .context(error::TokioPostgres)?;
+                    version += 1;
+                    // TODO log
+                    println!("Updated user database to schema version {}", version);
+                }
+                _ => return Ok(()),
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl UserDB for PostgresUserDB {
+    async fn register(&mut self, user: Validated<UserRegistration>) -> Result<UserId> {
+        let conn = self.conn_pool.get().await.map_err(|_| error::Error::BB8)?; // TODO: error reporting
+        let stmt = conn
+            .prepare(
+                "INSERT INTO users (id, email, password_hash, real_name, active) VALUES ($1, $2, $3, $4, $5);",
+            )
+            .await
+            .context(error::TokioPostgres)?;
+
+        let user = User::from(user.user_input);
+        conn.execute(
+            &stmt,
+            &[
+                &user.id.uuid(),
+                &user.email,
+                &user.password_hash,
+                &user.real_name,
+                &user.active,
+            ],
+        )
+        .await
+        .context(error::TokioPostgres)?;
+
+        Ok(user.id)
+    }
+
+    async fn login(&mut self, user: UserCredentials) -> Result<Session> {
+        todo!()
+    }
+
+    async fn logout(&mut self, session: SessionToken) -> Result<()> {
+        todo!()
+    }
+
+    async fn session(&self, token: SessionToken) -> Result<Session> {
+        todo!()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::util::user_input::UserInput;
+    use bb8_postgres::bb8::Pool;
+    use bb8_postgres::{tokio_postgres, PostgresConnectionManager};
+    use std::str::FromStr;
+
+    #[tokio::test]
+    async fn test() {
+        // TODO: load from test config
+        // TODO: add postgres to ci
+        // TODO: clear database
```

Comment:
Do we start Postgres here or do we mock the returned commands?

Reply:
maybe we don't test it at all. We have the In-Memory stores to test the rest of service for that reason that it's difficult to test with databases

Reply:
Can't we test it with a postgres configuration but only execute them with a special flag? We need some chance to test it I guess.

Reply:
currently the tests are ignore, the CI could also include ignore tests. However the more severe problem is that the rust test framework doesn't allow configuring running inidividual tests non-concurrently and also there is no way for setUp and tear down of tasks. So all tests would have to be in a single test function that creates the database in the beginning and drops it at the end. Right now I am testing user and project db separately but you have to clear the database manually and don't run them concurrently.

Reply:
Here is a quick workaround for setup and tear down:
https://medium.com/@ericdreichert/test-setup-and-teardown-in-rust-without-a-framework-ba32d97aa5ab

Maybe we can take a look at Diesel here how to do tests against PostgreSQL:
https://github.com/diesel-rs/diesel/blob/master/diesel/src/test_helpers.rs

Reply:
neat workaround. there is still the issue where tests may not run concurrently. Otherwise one test deletes the data that another test wants to check exists. We would need to have database transactions spanning multiple calls of our user/projectdb functions, but that's not possible right now and not necessary in practice. If we had this each test could just roll back the transaction at the end, leaving the database empty and other transactions undisturbed.

I don't think diesel solved this problem as they most likely test individual functionalities and not correct modifications of data stored in tables.

Reply:
Tests have to be complete and not depend on other tests.
If you want to have something in order, you should make one test out of it (that could call different methods).

You could lazy_static your instance in the test module and then acquire a mutex in the tests.

Reply:
its hard to make them independend if they all use the same data base. putting everything into one test seems more feasible. could we create a new task called testing databases where we tackle setting up postgres in the ci and combinding the tests. I would like to get this PR merged to continue working on other api changes.

Reply:
A task it is.